File: AN_2/SEM_2/MPP/c#/lab3/repository/ParticipantCursaRepository.cs
namespace lab3.repository;
using lab3.utils;
using lab3.domain;
using log4net;
using Microsoft.Data.Sqlite;

public class ParticipantCursaRepository : IParticipantCursaRepository
{
    private static readonly ILog logger = LogManager.GetLogger(typeof(ParticipantCursaRepository));
    private readonly DatabaseUtils dbUtils;

    public ParticipantCursaRepository(string connectionString)
    {
        dbUtils = new DatabaseUtils();
    }

    public List<ParticipantCursa> FindByCursa(Cursa cursa)
    {
        logger.Debug($"Incep cautarea participantilor pentru cursa cu id {cursa.GetId()}");
        List<ParticipantCursa> result = new List<ParticipantCursa>();
        string sql = @"SELECT p.idParticipant, p.nume, p.CNP, p.capacitate_motor, e.idEchipa, e.nume AS numeEchipa
                      FROM participant_cursa pc
                      JOIN participant p ON pc.idParticipant = p.idParticipant
                      JOIN echipa e ON p.idEchipa = e.idEchipa
                      WHERE pc.idCursa = @idCursa";

        try
        {
            using (var connection = dbUtils.GetConnection())
            using (var command = new SqliteCommand(sql, connection))
            {
                command.Parameters.AddWithValue("@idCursa", cursa.GetId());
                connection.Open();
                using (var reader = command.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        int participantId = reader.GetInt32(reader.GetOrdinal("idParticipant"));
                        string nume = reader.GetString(reader.GetOrdinal("nume"));
                        string CNP = reader.GetString(reader.GetOrdinal("CNP"));
                        int capacitateMotor = reader.GetInt32(reader.GetOrdinal("capacitate_motor"));
                        int echipaId = reader.GetInt32(reader.GetOrdinal("idEchipa"));
                        string numeEchipa = reader.GetString(reader.GetOrdinal("numeEchipa"));

                        Echipa echipa = new Echipa(echipaId, numeEchipa);
                        Participant participant = new Participant(participantId, nume, CNP, capacitateMotor, echipa);
                        result.Add(new ParticipantCursa(participant, cursa));
                    }
                }
            }

            logger.Debug($"Cautarea participantilor pentru cursa cu id {cursa.GetId()} s-a incheiat cu succes.");
        }
        catch (Exception e)
        {
            logger.Error($"Eroare la cautarea participantilor pentru cursa cu id {cursa.GetId()}", e);
        }

        return result;
    }

    public List<ParticipantCursa> FindByParticipant(Participant participant)
    {
        logger.Debug($"Incep cautarea curselor pentru participantul cu id {participant.GetId()}");
        List<ParticipantCursa> result = new List<ParticipantCursa>();
        string sql = @"SELECT c.idCursa, c.nume, c.capacitate_minima, c.capacitate_maxima
                      FROM participant_cursa pc
                      JOIN cursa c ON pc.idCursa = c.idCursa
                      WHERE pc.idParticipant = @idParticipant";

        try
        {
            using (var connection = dbUtils.GetConnection())
            using (var command = new SqliteCommand(sql, connection))
            {
                command.Parameters.AddWithValue("@idParticipant", participant.GetId());
                connection.Open();
                using (var reader = command.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        int cursaId = reader.GetInt32(reader.GetOrdinal("idCursa"));
                        string nume = reader.GetString(reader.GetOrdinal("nume"));
                        int capacitateMinima = reader.GetInt32(reader.GetOrdinal("capacitate_minima"));
                        int capacitateMaxima = reader.GetInt32(reader.GetOrdinal("capacitate_maxima"));

                        Cursa cursa = new Cursa(cursaId, nume, capacitateMinima, capacitateMaxima);
                        result.Add(new ParticipantCursa(participant, cursa));
                    }
                }
            }

            logger.Debug($"Cautarea curselor pentru participantul cu id {participant.GetId()} s-a incheiat cu succes.");
        }
        catch (Exception e)
        {
            logger.Error($"Eroare la cautarea curselor pentru participantul cu id {participant.GetId()}", e);
        }

        return result;
    }

    public void Adauga(ParticipantCursa participantCursa)
    {
        logger.Debug($"Incep adaugarea unui participant la cursa cu idCursa {participantCursa.GetCursa().GetId()}");
        string sql = "INSERT INTO participant_cursa (idParticipant, idCursa) VALUES (@idParticipant, @idCursa)";

        try
        {
            using (var connection = dbUtils.GetConnection())
            using (var command = new SqliteCommand(sql, connection))
            {
                command.Parameters.AddWithValue("@idParticipant", participantCursa.GetParticipant().GetId());
                command.Parameters.AddWithValue("@idCursa", participantCursa.GetCursa().GetId());
                connection.Open();
                command.ExecuteNonQuery();
            }

            logger.Debug(
                $"Adaugarea participantului cu id {participantCursa.GetParticipant().GetId()} la cursa cu idCursa {participantCursa.GetCursa().GetId()} s-a realizat cu succes.");
        }
        catch (Exception e)
        {
            logger.Error("Eroare la adaugarea participantului la cursa", e);
        }
    }

    public void Sterge(int participantId)
    {
        logger.Debug($"Incep stergerea participantului cu id {participantId}");
        string sql = "DELETE FROM participant_cursa WHERE idParticipant = @idParticipant";

        try
        {
            using (var connection = dbUtils.GetConnection())
            using (var command = new SqliteCommand(sql, connection))
            {
                command.Parameters.AddWithValue("@idParticipant", participantId);
                connection.Open();
                command.ExecuteNonQuery();
            }

            logger.Debug($"Stergerea participantului cu id {participantId} s-a realizat cu succes.");
        }
        catch (Exception e)
        {
            logger.Error($"Eroare la stergerea participantului cu id {participantId}", e);
        }
    }

    public void Modifica(ParticipantCursa participantCursa)
    {
        logger.Debug(
            $"Incep modificarea participantului cu idParticipant {participantCursa.GetParticipant().GetId()} pentru cursa cu idCursa {participantCursa.GetCursa().GetId()}");
        string sql = "UPDATE participant_cursa SET idCursa = @idCursa WHERE idParticipant = @idParticipant";

        try
        {
            using (var connection = dbUtils.GetConnection())
            using (var command = new SqliteCommand(sql, connection))
            {
                command.Parameters.AddWithValue("@idCursa", participantCursa.GetCursa().GetId());
                command.Parameters.AddWithValue("@idParticipant", participantCursa.GetParticipant().GetId());
                connection.Open();
                command.ExecuteNonQuery();
            }

            logger.Debug(
                $"Modificarea participantului cu idParticipant {participantCursa.GetParticipant().GetId()} pentru cursa cu idCursa {participantCursa.GetCursa().GetId()} s-a realizat cu succes.");
        }
        catch (Exception e)
        {
            logger.Error("Eroare la modificarea participantului cursa", e);
        }
    }

    public ParticipantCursa FindById(int id)
    {
        return null;
    }

    public List<ParticipantCursa> FindAll()
    {
        logger.Debug("Incep cautarea tuturor participantilor cursa");
        List<ParticipantCursa> result = new List<ParticipantCursa>();

        string sql = @"
                SELECT p.idParticipant, p.nume, p.CNP, p.capacitate_motor, 
                       e.idEchipa, e.nume AS echipaNume,
                       c.idCursa, c.nume AS cursaNume, c.capacitate_minima, c.capacitate_maxima
                FROM participant_cursa pc
                JOIN participant p ON pc.idParticipant = p.idParticipant
                JOIN echipa e ON p.idEchipa = e.idEchipa
                JOIN cursa c ON pc.idCursa = c.idCursa";

        try
        {
            using (var connection = dbUtils.GetConnection())
            using (var command = new SqliteCommand(sql, connection))
            {
                connection.Open();
                using (var reader = command.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        int participantId = reader.GetInt32(0);
                        string nume = reader.GetString(1);
                        string CNP = reader.GetString(2);
                        int capacitateMotor = reader.GetInt32(3);
                        int echipaId = reader.GetInt32(4);
                        string echipaNume = reader.GetString(5);
                        int cursaId = reader.GetInt32(6);
                        string cursaNume = reader.GetString(7);
                        int capacitateMinima = reader.GetInt32(8);
                        int capacitateMaxima = reader.GetInt32(9);

                        Echipa echipa = new Echipa(echipaId, echipaNume);
                        Participant participant = new Participant(participantId, nume, CNP, capacitateMotor, echipa);
                        Cursa cursa = new Cursa(cursaId, cursaNume, capacitateMinima, capacitateMaxima);
                        result.Add(new ParticipantCursa(participant, cursa));
                    }
                }
            }

            logger.Debug("Cautarea tuturor participantilor cursa s-a realizat cu succes.");
        }
        catch (Exception e)
        {
            logger.Error("Eroare la preluarea tuturor participantilor cursa", e);
        }

        return result;
    }
}

File: AN_2/SEM_2/MPP/java/lab6/Persistence/src/main/java/ro/mpp2024/persistence/respository/ParticipantCursaRepository.java
package ro.mpp2024.persistence.respository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.mpp2024.model.Cursa;
import ro.mpp2024.model.Echipa;
import ro.mpp2024.model.Participant;
import ro.mpp2024.model.ParticipantCursa;
import ro.mpp2024.persistence.ParticipantCursaRepositoryInterface;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ParticipantCursaRepository implements ParticipantCursaRepositoryInterface {
    private static final Logger logger = LogManager.getLogger(ParticipantCursaRepository.class);
    private final JdbcUtils dbUtils;

    public ParticipantCursaRepository(Properties props) {
        this.dbUtils = new JdbcUtils(props);
    }

    @Override
    public List<ParticipantCursa> findByCursa(Cursa cursa) {
        logger.traceEntry("Incep cautarea participantilor pentru cursa cu id " + cursa.getId());
        List<ParticipantCursa> result = new ArrayList<>();
        String sql = "SELECT p.idParticipant, p.nume, p.CNP, p.capacitate_motor, e.idEchipa, e.nume as echipaNume, c.idCursa, c.nume as cursaNume, c.capacitate_minima, c.capacitate_maxima " +
                "FROM participant_cursa pc " +
                "JOIN participant p ON pc.idParticipant = p.idParticipant " +
                "JOIN cursa c ON pc.idCursa = c.idCursa " +
                "JOIN echipa e ON p.idEchipa = e.idEchipa " +
                "WHERE pc.idCursa = ?";

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setInt(1, cursa.getId());
            ResultSet resultSet = statement.executeQuery();

            while (resultSet.next()) {
                int participantId = resultSet.getInt("idParticipant");
                String nume = resultSet.getString("nume");
                String CNP = resultSet.getString("CNP");
                int capacitateMotor = resultSet.getInt("capacitate_motor");
                int echipaId = resultSet.getInt("idEchipa");
                String echipaNume = resultSet.getString("echipaNume");
                int cursaId = resultSet.getInt("idCursa");
                String cursaNume = resultSet.getString("cursaNume");
                int capacitateMinima = resultSet.getInt("capacitate_minima");
                int capacitateMaxima = resultSet.getInt("capacitate_maxima");

                Echipa echipa = new Echipa(echipaId, echipaNume);
                Cursa cursaObj = new Cursa(cursaId, cursaNume, capacitateMinima, capacitateMaxima);
                Participant participant = new Participant(participantId, nume, CNP, capacitateMotor, echipa);

                result.add(new ParticipantCursa(participant, cursaObj));
            }

            logger.traceExit("Cautarea participantilor pentru cursa cu id " + cursa.getId() + " s-a incheiat cu succes.");
        } catch (SQLException e) {
            logger.error("Eroare la cautarea participantilor pentru cursa cu id " + cursa.getId(), e);
        }

        return result;
    }

    @Override
    public List<ParticipantCursa> findByParticipant(Participant participant) {
        logger.traceEntry("Incep cautarea curselor pentru participantul cu id " + participant.getId());
        List<ParticipantCursa> result = new ArrayList<>();
        String sql = "SELECT p.idParticipant, p.nume, p.CNP, p.capacitate_motor, e.idEchipa, e.nume as echipaNume, c.idCursa, c.nume as cursaNume, c.capacitate_minima, c.capacitate_maxima " +
                "FROM participant_cursa pc " +
                "JOIN participant p ON pc.idParticipant = p.idParticipant " +
                "JOIN cursa c ON pc.idCursa = c.idCursa " +
                "JOIN echipa e ON p.idEchipa = e.idEchipa " +
                "WHERE pc.idParticipant = ?";

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setInt(1, participant.getId());
            ResultSet resultSet = statement.executeQuery();

            while (resultSet.next()) {
                int participantId = resultSet.getInt("idParticipant");
                String nume = resultSet.getString("nume");
                String CNP = resultSet.getString("CNP");
                int capacitateMotor = resultSet.getInt("capacitate_motor");
                int echipaId = resultSet.getInt("idEchipa");
                String echipaNume = resultSet.getString("echipaNume");
                int cursaId = resultSet.getInt("idCursa");
                String cursaNume = resultSet.getString("cursaNume");
                int capacitateMinima = resultSet.getInt("capacitate_minima");
                int capacitateMaxima = resultSet.getInt("capacitate_maxima");

                Echipa echipa = new Echipa(echipaId, echipaNume);
                Cursa cursaObj = new Cursa(cursaId, cursaNume, capacitateMinima, capacitateMaxima);
                result.add(new ParticipantCursa(new Participant(participantId, nume, CNP, capacitateMotor, echipa), cursaObj));
            }

            logger.traceExit("Cautarea curselor pentru participantul cu id " + participant.getId() + " s-a incheiat cu succes.");
        } catch (SQLException e) {
            logger.error("Eroare la cautarea curselor pentru participantul cu id " + participant.getId(), e);
        }

        return result;
    }


    @Override
    public void adauga(ParticipantCursa ambele) {
        logger.traceEntry("Incep adaugarea unui participant la cursa cu idCursa " + ambele.getCursa().getId());
        String sql = "INSERT INTO participant_cursa (idParticipant, idCursa) VALUES (?, ?)";

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setInt(1, ambele.getParticipant().getId());
            statement.setInt(2, ambele.getCursa().getId());
            statement.executeUpdate();

            logger.traceExit("Adaugarea participantului cu idParticipant " + ambele.getParticipant().getId() + " la cursa cu idCursa " + ambele.getCursa().getId() + " s-a realizat cu succes.");
        } catch (SQLException e) {
            logger.error("Eroare la adaugarea participantului la cursa", e);
        }
    }

    @Override
    public void sterge(Integer id) {
        logger.traceEntry("Incep stergerea participantului cu id " + id);
        String sql = "DELETE FROM participant_cursa WHERE idParticipant = ?";

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setInt(1, id);
            statement.executeUpdate();

            logger.traceExit("Stergerea participantului cu id " + id + " s-a realizat cu succes.");
        } catch (SQLException e) {
            logger.error("Eroare la stergerea participantului cu id " + id, e);
        }
    }

    @Override
    public void modifica(ParticipantCursa ambele) {
        logger.traceEntry("Incep modificarea participantului cu idParticipant " + ambele.getParticipant().getId() + " pentru cursa cu idCursa " + ambele.getCursa().getId());
        String sql = "UPDATE participant_cursa SET idCursa = ? WHERE idParticipant = ?";

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setInt(1, ambele.getCursa().getId());
            statement.setInt(2, ambele.getParticipant().getId());
            statement.executeUpdate();

            logger.traceExit("Modificarea participantului cu idParticipant " + ambele.getParticipant().getId() + " pentru cursa cu idCursa " + ambele.getCursa().getId() + " s-a realizat cu succes.");
        } catch (SQLException e) {
            logger.error("Eroare la modificarea participantului cursa", e);
        }
    }

    @Override
    public ParticipantCursa findById(Integer id) {
        logger.traceEntry("Incep cautarea participantului cursa cu id " + id);
        return null;
    }

    @Override
    public List<ParticipantCursa> findAll() {
        logger.traceEntry("Incep cautarea tuturor participantilor cursa");
        List<ParticipantCursa> result = new ArrayList<>();
        String sql = "SELECT pc.idParticipant, p.nume, p.CNP, p.capacitate_motor, p.idEchipa, " +
                "c.idCursa, c.nume AS cursaNume, c.capacitate_minima, c.capacitate_maxima, " +
                "e.idEchipa, e.nume AS echipaNume " +
                "FROM participant_cursa pc " +
                "JOIN participant p ON pc.idParticipant = p.idParticipant " +
                "JOIN cursa c ON pc.idCursa = c.idCursa " +
                "LEFT JOIN echipa e ON p.idEchipa = e.idEchipa";

        try (Connection connection = dbUtils.getConnection();
             Statement statement = connection.createStatement();
             ResultSet resultSet = statement.executeQuery(sql)) {

            while (resultSet.next()) {
                int participantId = resultSet.getInt("idParticipant");
                String nume = resultSet.getString("nume");
                String CNP = resultSet.getString("CNP");
                int capacitateMotor = resultSet.getInt("capacitate_motor");
                int echipaId = resultSet.getInt("idEchipa");

                Echipa echipa = null;
                if (resultSet.wasNull()) {
                    logger.warn("Participant with idParticipant " + participantId + " does not have an associated echipa.");
                } else {
                    String echipaNume = resultSet.getString("echipaNume");
                    echipa = new Echipa(echipaId, echipaNume);
                }

                int cursaId = resultSet.getInt("idCursa");
                String cursaNume = resultSet.getString("cursaNume");
                int capacitateMinima = resultSet.getInt("capacitate_minima");
                int capacitateMaxima = resultSet.getInt("capacitate_maxima");
                Cursa cursaObj = new Cursa(cursaId, cursaNume, capacitateMinima, capacitateMaxima);

                Participant participant = new Participant(participantId, nume, CNP, capacitateMotor, echipa);
                result.add(new ParticipantCursa(participant, cursaObj));
            }

            logger.traceExit("Cautarea tuturor participantilor cursa s-a realizat cu succes.");
        } catch (SQLException e) {
            logger.error("Eroare la preluarea tuturor participantilor cursa", e);
        }

        return result;
    }

    @Override
    public int findNrPartByCursa(Cursa cursa) {
        logger.traceEntry("Incep cautarea participantilor pentru cursa cu id " + cursa.getId());
        int result = 0;
        String sql = "SELECT p.idParticipant, p.nume, p.CNP, p.capacitate_motor, e.idEchipa, e.nume as echipaNume, c.idCursa, c.nume as cursaNume, c.capacitate_minima, c.capacitate_maxima " +
                "FROM participant_cursa pc " +
                "JOIN participant p ON pc.idParticipant = p.idParticipant " +
                "JOIN cursa c ON pc.idCursa = c.idCursa " +
                "JOIN echipa e ON p.idEchipa = e.idEchipa " +
                "WHERE pc.idCursa = ?";

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setInt(1, cursa.getId());
            ResultSet resultSet = statement.executeQuery();

            while (resultSet.next()) {
                int participantId = resultSet.getInt("idParticipant");
                int cursaId = resultSet.getInt("idCursa");
                String cursaNume = resultSet.getString("cursaNume");
                int capacitateMinima = resultSet.getInt("capacitate_minima");
                int capacitateMaxima = resultSet.getInt("capacitate_maxima");

                result+=1;
            }

            logger.traceExit("Cautarea participantilor pentru cursa cu id " + cursa.getId() + " s-a incheiat cu succes.");
        } catch (SQLException e) {
            logger.error("Eroare la cautarea participantilor pentru cursa cu id " + cursa.getId(), e);
        }
        return result;
    }

    @Override
    public void adaugainbaza(ParticipantCursa participantCursa) {
        logger.traceEntry("Încep adăugarea în baza de date pentru ParticipantCursa: " +
                participantCursa.getParticipant().getId() + " + " +
                participantCursa.getCursa().getId());

        String sql = "INSERT INTO participant_cursa (idParticipant, idCursa) VALUES (?, ?)";

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setInt(1, participantCursa.getParticipant().getId());
            statement.setInt(2, participantCursa.getCursa().getId());
            statement.executeUpdate();

            logger.traceExit("Adăugarea s-a realizat cu succes.");
        } catch (SQLException e) {
            logger.error("Eroare la adăugarea ParticipantCursa în baza de date", e);
        }
    }

}
